### ISSUE TITLE
Hook in to Godot 4 POT generation

### ISSUE THREAD

**Issue by MarcusRiemer** [nathanhoad/godot_dialogue_manager#143]
**Is your feature request related to a problem? Please describe.**
I am not 100% sure how the workflow with `.po` files is intended to work and I think this might be due to a misunderstanding on either my side or your side. When I do translations for projects I build upon a single `.pot` file which contains no translations, but is effectively the "source" material. I go to "Godot Project Settings" => "Localization" => "POT Generation" and add all files that mark strings as being relevant for translation.

I then use [poEdit](https://poedit.net/) to create a file actually containing translations (`.po`) from the template file (`.pot`). If I update the source material, I can regenerate the `.pot` file from Godot and then the `.po` files via poEdit (see screenshot). 
![poEdit allows refreshing a .po file](https://user-images.githubusercontent.com/2138773/219748095-f5293959-86da-4682-aa83-fa0747b8de92.png)

This allows me to have a single `.po` file per language that I want to support, no matter how many source files "feed" translations into the `.pot` file.

With the solution implemented by Godot Dialogue Manager I'm not 100% certain what you would expect me to do once a `.dialogue` file has changed. The code seems like it keeps translations intact when re-exporting `.po` files, so I guess you wan't me to re-create the `.po` file for every `.dialogue` file that I have? This would mean that I need to work with quite the number of files (1 per language per level in my case) which doesn't seem very scalable. If I had only 4 scenes and 3 languages that's already 12 files I need to keep up with. And every time I change a dialogue file I need to re-export it 3 times (once per language).

**Describe the solution you'd like**
The most beautiful solution for me would be if you could hook into the Godot Translation system so that I could add .dialogue files immediately like any other source of translatable content. In the screenshot I have just added the file, even though Godot obviously can't use it yet.

![Add .dialogue files here](https://user-images.githubusercontent.com/2138773/219749260-a7929b16-0899-4690-85d6-8183dcc28f7a.png)

**Describe alternatives you've considered**
Alternatively I would love to have a single button in the Dialogue Plugin that allows me to export all translatable strings of `.dialogue` files. Either way I would leave the difficult part of updating `.po` files to the specialized translation tools.

**Additional context**
* From my PoV generating `.pot` files and leaving the updating part should make the code base easier over all. I can't make any promises, but I will ping you back if I should find time to take a look at it myself.
* Many Kudos for providing a way to mark strings with a translation ID! I love the fact that I can continue to write the `.dialogue` files in plain english and don't have to fear re-matching translation strings once I make a change in the source material.

**Comment by nathanhoad** [comment]
I admittedly don't use the PO exports (it was contributed by @Festerdam) in my own project but I'll have a look at your suggestion. Do you have some example files you could provide so I know what the exports should look like?

**Comment by nathanhoad** [comment]
I've added a thing that hooks into that "Generate POT" button. Let me know if that's working how you'd expect.

I'm also wondering if that's a better solution than having the PO export in the Dialogue Manager UI?

**Comment by Festerdam** [comment]
Hi, I'm the human that initially made the PR allowing PO generation (https://github.com/nathanhoad/godot_dialogue_manager/pull/64). It was far from an ideal solution, but it allowed one to generate a PO file. This PO file could then be used to generate a POT file (and from that the remaining PO files for the other languages). In my project I use the following Python script to do that:
```
#!/bin/python3
#     FCT RPG
#     Copyright (C) 2022  Núcleo de Design e Desenvolvimento de Videojogos
#
#     This program is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with this program.  If not, see <https://www.gnu.org/licenses/>.

LOCALES = ["en_US", "pt_PT", "de_DE"]
DIALOGUE_PATH = "src/client/dialogue"

import sys
from pathlib import Path
from subprocess import run

def generate_pot(proj_root = ".."):
    run(["pybabel", "extract", "-F", "locale/babelrc", "-k", "text", "-k", "LineEdit/placeholder_text", "-k", "tr", "-o", "locale/messages.pot", "src"], cwd=proj_root)
    dialogue_path = Path(proj_root + "/" + DIALOGUE_PATH)
    for f in dialogue_path.glob("*.po"):
        run(["msgfilter", "-i", str(f.relative_to(proj_root)), "-o", "locale/" + f.with_suffix(".pot").with_stem(f.stem.rsplit("-", maxsplit=1)[0]).name, "--keep-header", "true"], cwd=proj_root)
        run("cp " + str(f.relative_to(proj_root)).replace(".po", ".*") + " locale/", cwd=proj_root, shell=True)

def generate_po(proj_root = ".."):
    p = Path(proj_root + "/locale/")
    for pot in p.glob("*.pot"):
        for locale in LOCALES:
            f = pot.with_stem(pot.stem + "-" + locale).with_suffix(".po")
            if f.is_file():
                run(["msgmerge", "--update", "--backup=none", str(f.relative_to(proj_root)), str(pot.relative_to(proj_root))], cwd=proj_root)
            else:
                run(["msginit", "--no-translator", "--input=" + str(pot.relative_to(proj_root)), "--locale=" + locale, "-o", str(f.relative_to(proj_root))], cwd=proj_root)

if __name__ == "__main__":
    if len(sys.argv) == 3:
        if sys.argv[1] == "generate_pot":
            generate_pot(sys.argv[2])
        if sys.argv[1] == "generate_po":
            generate_po(sys.argv[2])
        if sys.argv[1] == "all":
            generate_pot(sys.argv[2])
            generate_po(sys.argv[2])
    elif len(sys.argv) == 2:
        if sys.argv[1] == "generate_pot":
            generate_pot()
        if sys.argv[1] == "generate_po":
            generate_po()
        if sys.argv[1] == "all":
            generate_pot()
            generate_po()
    else:
        print("Usage: ./update_localization.py {generate_pot|generate_po|all } [proj_root]")
```
Where all the PO and POT files are saved under `locale`, the dialogue files under `src/client/dialogue` and the script is saved unter `utils`.

So the usual workflow for me was to write some dialogue, save the dialogue, save it again to a PO file, run `./update_localization.py all` within the `utils` directory, and then with a PO editor add the translations for the new strings in the updated PO files of the other languages.

Truly, it isn't really useful to export a single PO file, without additional code. That feature was useful for me since I was willing to write a script to automatically generate a pot file from the po file and then generate the remaining po files, but I thought that my code could be useful to other people regardless, that's why I sent in the PR. I should have clarified that in my PR and I'm sorry I didn't.

**Comment by nathanhoad** [comment]
Thanks for shedding some extra light on this. 

Does the new POT generation work better for you as well? If so, I'd like to remove the individual PO generation to cut back on complexity and in favour of this newer stuff.

**Comment by nathanhoad** [comment]
@MarcusRiemer All dialogue files will now be automatically added to the POT Generation file list. I've also updated the Point n Click example scene to use a `po` translation for the German (which tells me that it at least works in a basic sense).

I'll close this issue as complete but feel free to reopen if something comes up.

**Comment by Festerdam** [comment]
> Thanks for shedding some extra light on this.
> 
> Does the new POT generation work better for you as well? If so, I'd like to remove the individual PO generation to cut back on complexity and in favour of this newer stuff.

Works for me.

**Comment by MarcusRiemer** [comment]
Wow, the children kept me busy over the weekend and now I am waking up to this beautiful surprise. Thanks alot, this really improves the translation experience by a wide margin. I really didn't expect such a quick response and I really appreciate it.
 
There is still something ... off ... though with the manner in which IDs and text are presented, and I am not 100% sure whether this is a shortcoming of poEdit or the generated file. As the text to translate is added as a comment and the ID is the text itself, it a) throws all kind of false positives for punctuation and b) requires one to copy the comments manually for the default language.

![image](https://user-images.githubusercontent.com/2138773/220058794-75f9c800-4f30-4921-a8ba-7aee15fa4ed4.png)

This really might be an issue with poEdit though. It could just be that this is a logical consequence of (imho rightfully) using unique IDs to keep track of texts. I will investigate and come back.

**Comment by nathanhoad** [comment]
Yeah, from what I can tell poEdit isn't expecting actual IDs for the `msgid`. I did try swapping the ID and the context but it (expectedly) doesn't update properly.

**Comment by MarcusRiemer** [comment]
Ultimately it seems that poEdit just follows a different design philosophy and very strongly prefers the message IDs to be in the root language. I didn't stumble over this in my previous years of working with it because I always had semi-readable message ids like "region_a_person_b_greeting".

For the moment I am just moving on without artificial message IDs and poEdit does a surprisingly good job to match updated message IDs in case of minor changes.
